pkg/worker/log/logger.go

```
@@ -0,0 +1,114 @@
+		watcher: NewWatcher(),
+	}
+}
+
+// Path returns a absolute file path given a name.
+func (l *Logger) Path(name string) string {
+	return filepath.Join(l.config.LogFolder, fmt.Sprintf("%s.log", name))
+}
+
+// Create creates and return a os.File under log folder.
+// If the file can't be created an error will be returned.
+func (l *Logger) Create(name string) (*os.File, error) {
+	return os.Create(l.Path(name))
+}
+
+// Remove removes the named file under the log folder.
+func (l *Logger) Remove(name string) error {
+	return os.Remove(l.Path(name))
+}
+
+// Tailf watch a named log file under the log folder, and
+// streams his content through a channel.
+// If the context is canceled, the tail will stop to watch the
+// log file, and the channel will be closed.
+func (l *Logger) Tailf(ctx context.Context, name string) (chan string, error) {
+	file, err := os.OpenFile(l.Path(name), os.O_RDONLY, 0644)
+	if os.IsNotExist(err) {
+		return nil, err
+	}
+	// watching for modify and delete events
+	eventchan, err := l.watcher.Watch(ctx, l.Path(name), syscall.IN_MODIFY|syscall.IN_DELETE)
+	if err != nil {
+		file.Close()
+		return nil, err
+	}
+	logchan := make(chan string)
+	go func() {
+		defer func() {
+			file.Close()
+			close(logchan)
+		}()
+		reader := bufio.NewReader(file)
+		offset := 0
+		// stream the file from the begin
+		for {
+			// read line by line
+			bytesr, err := reader.ReadBytes('\n')
+			if err != nil {
+				break
+			}
+			select {
+			case logchan <- string(bytesr):
+			case <-ctx.Done():
+				return
+			}
+			// calculate the offset
+			offset += len(bytesr)
+		}
+		// handle file events to keep reading
+		for {
+			select {
+			case event := <-eventchan:
+				// if the file was written the content should be streamed
+				if event.Modified() {
+					// stream the file from the last offset
+					file.Seek(int64(offset), 0)
+					for {
+						// read line by line
+						bytesr, err := reader.ReadBytes('\n')
+						if err != nil {
```

Information about this error is lost.